### PR TITLE
[wicketd] Respect installinator's report of which disks were written

### PR DESCRIPTION
**Pull request by jgallagher**
We use this in two (new) ways:

1. We write the corresponding host boot flash slots (previously we only and always wrote slot 0)
2. Set the host to boot off of "the first disk successfully written" (previously we didn't set this at all, and left the host booting off of slot 0 as a side effect of trampoline delivery)

A big chunk of the diff is shifting engine steps to deliver an update to a component via the SP into a new nested engine, making it easier to loop over the disks returned by installinator (which are only available to us inside a running step).

### CODE REVIEW

File: wicketd/src/update_tracker.rs [hunk 1]
```diff
@@ -1134,14 +1089,61 @@ impl UpdateContext {
...
+                    };
+
+                    match metadata {
+                        InstallinatorCompletionMetadata::Write { output } => {
+                            write_output = Some(output.clone());
+                        }
+                        InstallinatorCompletionMetadata::HardwareScan { .. }
+                        | InstallinatorCompletionMetadata::ControlPlaneZones { .. }
+                        | InstallinatorCompletionMetadata::Download { .. }
+                        | InstallinatorCompletionMetadata::Unknown => (),
```
Comment (sunshowers) [inline wicketd/src/update_tracker.rs:1135]: Do we need to express all the use cases here?

File: wicketd/src/update_tracker.rs [hunk 2]
```diff
@@ -1134,14 +1089,61 @@ impl UpdateContext {
...
+                            ..
+                        } => outcome,
+                        StepEventKind::NoStepsDefined
+                        | StepEventKind::ExecutionStarted { .. }
+                        | StepEventKind::ProgressReset { .. }
+                        | StepEventKind::AttemptRetry { .. }
+                        | StepEventKind::ExecutionFailed { .. }
+                        | StepEventKind::Nested { .. }
+                        | StepEventKind::Unknown => continue,
+                    };
```
Comment (sunshowers) [inline wicketd/src/update_tracker.rs:1118]: Wondering how you would feel about making this a method on `StepEventKind`.

File: wicketd/src/update_tracker.rs [hunk 3]
```diff
@@ -1134,14 +1089,61 @@ impl UpdateContext {
...
+
+                    // We only care about successful (including "success with
+                    // warning") outcomes.
+                    let metadata = match outcome {
+                        StepOutcome::Success { metadata }
+                        | StepOutcome::Warning { metadata, .. } => metadata,
+                        StepOutcome::Skipped { .. } => continue,
+                    };
```
Comment (sunshowers) [inline wicketd/src/update_tracker.rs:1126]: And this a method on `StepOutcome` (`completion_metadata`). That will avoid a lot of this boilerplate.

**Review by sunshowers — APPROVED** [review]
Looks good, just a couple minor comments that I'm fine resolving either way.

File: wicketd/src/update_tracker.rs (hunk 2 above)
Comment (jgallagher) [reply wicketd/src/update_tracker.rs]: Yeah, seems perfectly reasonable. Moved in 7c6dceef

File: wicketd/src/update_tracker.rs (hunk 3 above)
Comment (jgallagher) [reply wicketd/src/update_tracker.rs]: Yeah, love that. Done in 7c6dceef

**Pull request merged**